### ISSUE TITLE
Hanging Validation

### ISSUE THREAD

**Issue by wallarug** [pytorch/pytorch#72897]
### 🐛 Describe the bug

When running validation in my code, it sometimes hangs in the latest nightly release of pytorch version `1.12.0.dev20220215+cu113`.  This usually occurs after 1 or 2 epochs.   

When testing against last month's latest nightly release, there were no issues.  I managed to run the code for more than 300 epochs unimpeded.     The working version was pytorch version `1.11.0.dev20220117+cu113` works fine.

This is a sub-section of the code, with the validation phase.

```python3
    if rank == 0:
            with torch.no_grad():
                model.eval()
                for i, sample in enumerate(valid_loader):

                    images = sample['data'].cuda(non_blocking=True)
                    labels = sample['label'].cuda(non_blocking=True)
                    weights = sample['weight'].cuda(non_blocking=True)

                    with torch.cuda.amp.autocast():
                        outputs = model.module(images)
                        pred = outputs * weights
                        loss = loss_fn(pred, labels)
                print("validation loss: " + str(loss.mean(dim=0).sum().item()))        
                model.train()
        dist.barrier()
```

Output:
No error message, just hangs

Interestingly, if you move the print block under the "for block" it runs fine.

I have seen numerous issues similar to this from March 2021, however, they seem to have been resolved.  This seems like a new issue.

### Versions

```
Collecting environment information...
PyTorch version: 1.12.0.dev20220215+cu113
Is debug build: False
CUDA used to build PyTorch: 11.3
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.8.10 (default, Nov 26 2021, 20:14:08)  [GCC 9.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-1027-aws-x86_64-with-glibc2.29
Is CUDA available: True
CUDA runtime version: 11.3.109

Nvidia driver version: 470.82.01
cuDNN version: Probably one of the following:
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn.so.8.3.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.3.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.3.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.3.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.3.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.3.1
/usr/local/cuda-11.3/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.3.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.22.1
[pip3] torch==1.12.0.dev20220215+cu113
[pip3] torchaudio==0.11.0.dev20220215+cu113
[pip3] torchvision==0.13.0.dev20220215+cu113
[conda] Could not collect
```

Any assistance on this would be great.

**Comment by zou3519** [comment]
A hang sounds concerning. Do you have a script we can use to reproduce the problem?

**Comment by wallarug** [comment]
Hey @zou3519 ,

I have an update on this one.  Seems to hang in 1.11.0 version as well when you scale up to more machines.  Distributed across gpus runs fine.  When you scale up to say 4 machines, it doesn't work anymore and just hangs again.

I can't provide the full script (at least publicly on GitHub) to replicate the issue as it is part of a company/work contract protected by IP agreements (hope you understand).  It's also a bit complex and dependant on internal code/data structures.

I followed the PyTorch Distributed tutorial documentation, am using the nccl backend and have used a custom dataset structure.

Everything works fine when I comment out the validation / don't run it. 

Happy to provide more code samples offline/via email that may help identify and solve this problem.

**Comment by zou3519** [comment]
Moving this issue over to the Distributed queue, hopefully one of the Distributed Team folks can help.

**Comment by cbalioglu** [comment]
Just to clarify. You observe this issue only when the line
```
print("validation loss: " + str(loss.mean(dim=0).sum().item()))
```
is in your script?

**Comment by wallarug** [comment]
Yep.  I haven't been able to replicate without the print statement.

I suspect it is actually the `loss.mean(dim=0).sum().item()` that is causing the issue.

**Comment by cbalioglu** [comment]
cc @zhaojuanmao

**Comment by zhaojuanmao** [comment]
The validation codes should not have any communication, so it should not hang at all. I guess it hangs at dist.barrier(), mostly means some ranks never arrive at dist.barrier() for some reasons in your application

**Comment by wallarug** [comment]
I was never able to fix this issue.  I ended up just doing validation on rank 0 without any DDP.

**Comment by EphChem** [comment]
+1, encountering exactly the same issue.  Any progress on this?

**Comment by papy-ganjha** [comment]
There is no current fix for this issue. But it will be solved in the folowing realeases of pytorch. To solve this you can use the nigthly build of pytorch. 
See this issue which is related I think of what you want. #54059

**Comment by wanchaol** [comment]
Let us know if latest nightly still doesn't work
